dotnet/runtime
# [browser][file system] Tests System.IO.FileSystem

Added 45 `[PlatformSpecific(~TestPlatforms.Browser)]` checks to tests.

```
 info: test[0]
        Tests run: 3804, Errors: 0, Failures: 0, Skipped: 100. Time: 37.597637s

```

```
// IO.Pipes not supported
public void FlushCanBeUsedOnPipes_Success(bool? flushToDisk)
public void CanSeekReturnsFalseForPipe()
public async Task AnonymousPipeReadViaFileStream(bool asyncReads)
public async Task AnonymousPipeWriteViaFileStream(bool asyncWrites)
public async Task AnonymousPipeViaFileStream_AllDataCopied(int writeSize, int numWrites)

// System.PlatformNotSupportedException : Cannot wait on monitors on this runtime.
public void EmptyFile_Read_Succeeds()
public void NonEmptyFile_Read_GetsExpectedData()
public void NonEmptyWrite_WritesExpectedData()
public Task ManyConcurrentWriteAsyncs()

```

- Directory.CreateDirectory.cs

   - Browser platform volume does not limit each component of the path to a total of 255 characters.
   - Remove the test `DirectoryWithComponentLongerThanMaxComponentAsPath_ThrowsException` for Browser platform.
   - Add new test `DirectoryWithComponentLongerThanMaxComponentAsPath_BrowserDoesNotThrowsException` for Browser platform in case this check is added in the future.


~~Error number:~~  https://github.com/dotnet/runtime/issues/39955 - Resolved

File locking not supported: https://github.com/dotnet/runtime/issues/40065

```
System.IO.FileSystem/tests/File/ReadWriteAllBytes.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public void OpenFile_ThrowsIOException()

System.IO.FileSystem/tests/File/ReadWriteAllBytesAsync.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public async Task OpenFile_ThrowsIOExceptionAsync()

System.IO.FileSystem/tests/File/ReadWriteAllLines.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public void OpenFile_ThrowsIOException()

        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public void OpenFile_ThrowsIOException()

System.IO.FileSystem/tests/File/ReadWriteAllLinesAsync.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public async Task OpenFile_ThrowsIOExceptionAsync()

System.IO.FileSystem/tests/File/ReadWriteAllText.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public void OpenFile_ThrowsIOException()

System.IO.FileSystem/tests/File/ReadWriteAllTextAsync.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public async Task OpenFile_ThrowsIOExceptionAsync()

System.IO.FileSystem/tests/File/Create.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public void FileInUse()

System.IO.FileSystem/tests/FileStream/ctor_str_fm_fa_fs.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public void NoTruncateOnFileShareViolation(FileMode fileMode)

System.IO.FileSystem/tests/FileStream/ctor_str_fm_fa_fs.write.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public void FileShareWithoutWriteThrows()
```

A FIFO special file is similar to a pipe, except that it is created in a different way. Instead of being an anonymous communications channel, a FIFO special file is entered into the file system by calling mkfifo().  Pipes are not supported on WebAssembly.  Platform specific skip for browser in the following tests.

```
System.IO.FileSystem/tests/DirectoryInfo/Exists.cs
        public void FalseForNonRegularFile()

System.IO.FileSystem/tests/Directory/Exists.cs
        public void FalseForNonRegularFile()

System.IO.FileSystem/tests/FileInfo/Exists.cs
        public void TrueForNonRegularFile()

System.IO.FileSystem/tests/File/Exists.cs
        public void FalseForNonRegularFile()
```

~~File rename issue reporting wrong error:~~  https://github.com/dotnet/runtime/issues/40305
**_PR #40310 works around the issue for now while waiting for fix emscripten-core/emscripten#11804 / emscripten-core/emscripten#11812_**

~~Depends on PR: https://github.com/dotnet/runtime/pull/39763~~ - merged
